### PR TITLE
e2e test transparent SA rewrite in downstream deployment

### LINKED ISSUE #758: Write missing e2e tests for the transparent SA rewrite in downstream deployments functionality.
The functionality added on https://github.com/kcp-dev/kcp/pull/670 is missing e2e tests.

### PR DESCRIPTION
**Pull request by geoberle**
## Summary

deploy a demo application via KCP to a downstream cluster. this application explicitly uses the in-cluster service account rewrite to create a configmap in an upstream KCP namespace. the e2e test verifies, that this configmap gets created.

this part of the e2e test only runs when a workload cluster is available, e.g. during CI or when the testflags `--pcluster-kubeconfig`and `--kcp-test-image` are present. the GH-action CI provides a freshly build in-cluster test image automatically. a prebuild multi-arch image is available under ghcr.io/kcp-dev/kcp/kcp-test-image:main

## Related issue(s)

Fixes #758

Signed-off-by: Gerd Oberlechner <goberlec@redhat.com>

### CODE REVIEW

**Comment by geoberle** [comment]
> This looks great! I imagine when we put the correct namespace into the in-cluster configuration (as opposed to the downward-api based one in there now) you won't need to pass the NS in.

Yes absolutely! https://github.com/kcp-dev/kcp/pull/1033

File: .github/workflows/icc-test-image.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+name: Build in-cluster configuration test Image
```
Comment (sttts) [inline .github/workflows/icc-test-image.yaml:1]: we will get many more tests that need an image. Worth to make this generic and have multple binaries in it?

**Comment by sttts** [comment]
:+1:

File: test/e2e/fixtures/icc-test-app/icc-test-app.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	"os"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+func main() {
+	namespaceName := os.Getenv("TARGET_NAMESPACE")
```
[the remark this answers is not in the document]
Comment (sttts) [reply test/e2e/fixtures/icc-test-app/icc-test-app.go:31]: xref https://github.com/kcp-dev/kcp/pull/1033

File: test/e2e/syncer/syncer_test.go [hunk 3]
```diff
@@ -27,6 +27,7 @@ import (
 
 	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/rbac/v1"
```
[the remark this answers is not in the document]
Comment (geoberle) [reply test/e2e/syncer/syncer_test.go]: [6cbd4d9](https://github.com/kcp-dev/kcp/pull/1044/commits/6cbd4d90d36f007938aba773c33d06e29a8062b8)

File: test/e2e/fixtures/icc-test-app/icc-test-app.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+		log.Panicln("failed to create in-cluster config", err)
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Panicln("failed to create clientset with in-cluster config", config, err)
+	}
+
+	configMap := createConfigMap(namespaceName, configMapName)
+	_, err = clientset.CoreV1().ConfigMaps(namespaceName).Create(context.TODO(), configMap, metav1.CreateOptions{})
```
[the remark this answers is not in the document]
Comment (geoberle) [reply test/e2e/fixtures/icc-test-app/icc-test-app.go]: [4252b4f](https://github.com/kcp-dev/kcp/pull/1044/commits/4252b4fa48782930b189c54ecd1edd39a7a5ee53)

File: test/e2e/fixtures/icc-test-app/icc-test-app.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	if err != nil {
+		log.Panicln("failed to create configmap", err)
+	}
+
+	log.Printf("configmap %s created in namespace %s\n", namespaceName, configMap)
+}
+
+func createConfigMap(namespace string, name string) *corev1.ConfigMap {
```
[the remark this answers is not in the document]
Comment (geoberle) [reply test/e2e/fixtures/icc-test-app/icc-test-app.go]: i inlined it in [4252b4f](https://github.com/kcp-dev/kcp/pull/1044/commits/4252b4fa48782930b189c54ecd1edd39a7a5ee53)

File: test/e2e/syncer/syncer_test.go [hunk 6]
```diff
@@ -167,9 +168,102 @@ func TestSyncerLifecycle(t *testing.T) {
...
+					Name: "default",
+				},
+			},
+		}, metav1.CreateOptions{})
+		require.NoError(t, err, "failed to create upstream rolebinding")
+
+		t.Logf("Creating upstream in-cluster configuration test deployment")
+		expectedConfigMapName := "expected-configmap"
+		iccUpstreamDeployment, err := upstreamKubeClient.AppsV1().Deployments(upstreamNamespace.Name).Create(ctx, &appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
```
[the remark this answers is not in the document]
Comment (geoberle) [reply test/e2e/syncer/syncer_test.go:209]: you mean like `embeddedResources.ReadFile("deployment.yaml")` for the icc test deployment and the RBAC objects?

File: .github/workflows/icc-test-image.yaml (hunk 1 above)
Comment (ncdc) [reply .github/workflows/icc-test-image.yaml:1]: Maybe we start by renaming this from icc-test-image to kcp-test-image (or something more generic like that), but save multiple binaries/multiple subcommands for a follow-up when we have a need for them?

File: test/e2e/syncer/syncer_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (geoberle) [reply test/e2e/syncer/syncer_test.go]: [29e4bd8](https://github.com/kcp-dev/kcp/pull/1044/commits/29e4bd8047b70bc41357b213a727e9c8ce49fb9e)

**Review by stevekuznetsov — APPROVED** [review]

**Pull request merged**